samza-kafka/src/main/scala/org/apache/samza/system/kafka/NewKafkaSystemConsumer.java

```diff
@@ -0,0 +1,412 @@
+      return;
+    }
+    super.register(systemStreamPartition, offset);
+
+    TopicPartition tp = toTopicPartition(systemStreamPartition);
+
+    topicPartitions2SSP.put(tp, systemStreamPartition);
+
+    LOG.info("Registering ssp = " + systemStreamPartition + " with offset " + offset);
+
+    String existingOffset = topicPartitions2Offset.get(tp);
+    // register the older (of the two) offset in the consumer, to guarantee we do not miss any messages.
+    if (existingOffset == null || compareOffsets(existingOffset, offset) > 0) {
+      topicPartitions2Offset.put(tp, offset);
+    }
+
+    samzaConsumerMetrics.registerTopicAndPartition(toTopicAndPartition(tp));
+  }
+
+  /**
+   * Compare two String offsets.
+   * Note. There is a method in KafkaAdmin that does that, but that would require instantiation of systemadmin for each consumer.
+   * @param off1
+   * @param off2
+   * @return see {@link Long#compareTo(Long)}
+   */
+  public static int compareOffsets(String off1, String off2) {
+    return Long.valueOf(off1).compareTo(Long.valueOf(off2));
+  }
+
+  @Override
+  public String toString() {
+    return systemName + "/" + clientId + "/" + super.toString();
+  }
+
+  @Override
+  public Map<SystemStreamPartition, List<IncomingMessageEnvelope>> poll(
+      Set<SystemStreamPartition> systemStreamPartitions, long timeout) throws InterruptedException {
+
+    // check if the proxy is running
+    if (!proxy.isRunning()) {
+      stop();
+      if (proxy.getFailureCause() != null) {
+        String message = "KafkaConsumerProxy has stopped";
+        throw new SamzaException(message, proxy.getFailureCause());
+      } else {
+        LOG.warn("Failure cause is not populated for KafkaConsumerProxy");
+        throw new SamzaException("KafkaConsumerProxy has stopped");
+      }
+    }
+
+    Map<SystemStreamPartition, List<IncomingMessageEnvelope>> res = super.poll(systemStreamPartitions, timeout);
+    return res;
+  }
+
+  public static TopicAndPartition toTopicAndPartition(TopicPartition tp) {
+    return new TopicAndPartition(tp.topic(), tp.partition());
+  }
+
+  public static TopicAndPartition toTopicAndPartition(SystemStreamPartition ssp) {
+    return new TopicAndPartition(ssp.getStream(), ssp.getPartition().getPartitionId());
+  }
+
+  public static TopicPartition toTopicPartition(SystemStreamPartition ssp) {
+    return new TopicPartition(ssp.getStream(), ssp.getPartition().getPartitionId());
+  }
+
+  public static SystemStreamPartition toSystemStreamPartition(String systemName, TopicAndPartition tp) {
+    return new SystemStreamPartition(systemName, tp.topic(), new Partition(tp.partition()));
+  }
+
+  /**
+   * return system name for this consumer
+   * @return system name
+   */
+  public String getSystemName() {
+    return systemName;
+  }
+
+  ////////////////////////////////////
+  // inner class for the message sink
+  ////////////////////////////////////
```

Comment:
Correct me if i'm wrong here. This appears to be used only within the `KafkaConsumerProxy`. Can we make it an internal static class of the `KafkaConsumerProxy` class instead?